### PR TITLE
Zippy_scan closes #1951

### LINKED ISSUE #1951: [Analyzer] Zippy
## Name
Zippy

## Link
https://github.com/thinkst/zippy

## Type of analyzer
file analyzer

## Why should we use it
Detect AI-generated text [relatively] quickly via compression ratios

## Possible implementation
tool written in python, we could integrate it as a library

### PR DESCRIPTION
**Pull request by g4ze**
close #1951

# Description

implement Zippy_scan
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [x] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.
image
raw json:
{0:"AI",
1:0.09720537891081449,
2:"Abstract While large-scale pretrained language models have been shown to learn effective linguistic representations for many NLP tasks, there remain many real-world contextual aspects of language that current approaches do not capture. For instance, consider a clozetest I enjoyed the game this weekend: the correct answer depends heavily on where the speaker is from, when the utterance occurred, and the speakerÃÂs broader social milieu and preferences. Although language depends heavily on the geographical, temporal, and other social contexts of the speaker, these elements have not been incorporated into modern transformer-based language models. We propose a simple but effective approach to incorporate speaker social context into the learned representations of large-scale language models. Our method first learns dense representations of social contexts using graph representation learning algorithms and then primes language model pretraining with these social context representations. We evaluate our approach on geographically-sensitive languagemodeling tasks and show a substantial improvement (more than 100% relative lift on MRR) compared to baselines. Introduction Language models are at the very heart of many modern NLP systems and applications (Young et al., 2018). Representations derived from largescale language models are used widely in many downstream NLP models (Peters et al., 2018; Devlin et al., 2019). However, an implicit assumption made in most modern NLP systems (including language models) is that language is independent of extra-linguistic context such as speaker/author identity and their social setting. While this simplifying assumption has undoubtedly encouraged remarkable progress in modeling language, there is overwhelming evidence in socio-linguistics that language understanding is influenced by the social context in which language is grounded (Nguyen et al., 2016; Hovy, 2018; Mishra et al., 2018; Garten et al., 2019; Flek, 2020; Bender and Koller, 2020). In fact, language use on social media where every utterance is grounded in a specific social context (like time, geography, social groups, communities) reinforces this often ignored aspect of language. When NLP applications ignore this social context, they may perform sub-optimally underscoring the need for a richer integration of social contexts into NLP models (Pavalanathan et al., 2015; Lynn et al., 2017; Zamani et al., 2018; Lynn et al., 2019; May et al., 2019; Kurita et al., "}

### CODE REVIEW

**Comment by g4ze** [comment]
@mlodic pls review

**Comment by g4ze** [comment]
i read the analyzer test classes and the `no job found` error is probably because we don't have any `.txt` file in our `test_files`, the the test fails as it is not able to find any matching file for the supported file type in the `mimetype`. We do have an option to select `test/plain` as supported filetype in django admin but apparently no tests cover that. This might be a missing edge case which might have been overlooked while writing test? I can be completely wrong with that tho

**Review by mlodic — CHANGES_REQUESTED** [review]
please refine the code, it seems still WIP.

Plus, read the checklist completely, it's there to be read. Thanks

**Comment by g4ze** [comment]
apologies for the ignorance in reading the PR checklist, my bad.

File: api_app/analyzers_manager/file_analyzers/zippy_scan.py [hunk 1]
```diff
@@ -17,31 +17,36 @@ class ZippyAnalyser(FileAnalyzer):
...
-            z = Zippy(engine=CompressionEngine.BROTLI)
-        else:
-            z = EnsembledZippy()
+        z = (
+            Zippy(engine=self.ENGINES[self.engine])
+            if self.engine in self.ENGINES
+            else EnsembledZippy()
+        )
 
         logger.info("Running Zippy on file %s using %s", self.filepath, self.engine)
```
Comment (mlodic) [inline api_app/analyzers_manager/file_analyzers/zippy_scan.py:38]: may I ask you to use f-strings instead of %s?

File: api_app/analyzers_manager/file_analyzers/zippy_scan.py [hunk 2]
```diff
@@ -17,31 +17,36 @@ class ZippyAnalyser(FileAnalyzer):
...
         try:
-            response = z.run_on_file_chunked(filename)
-        except Exception:
+            text_data = binary_data.decode("utf-8")
+            response = z.run_on_file_chunked(filename=self.filepath)
+        except UnicodeDecodeError:
+            logger.exception("Cannot decode file %s", self.filepath)
+            raise AnalyzerRunException("Cannot decode file")
+        except Exception as e:
             logger.exception("%s.run_on_text_chunked(text_data) failed", self.engine)
```
Comment (mlodic) [inline api_app/analyzers_manager/file_analyzers/zippy_scan.py:47]: same here

File: api_app/analyzers_manager/file_analyzers/zippy_scan.py [hunk 3]
```diff
@@ -17,31 +17,36 @@ class ZippyAnalyser(FileAnalyzer):
...
-            raise AnalyzerRunException
+            logger.exception(e)
```
Comment (mlodic) [inline api_app/analyzers_manager/file_analyzers/zippy_scan.py:48]: can you merge the 2 exception logs together in a single string? this is helpful once you need to debug the app

File: tests/api_app/analyzers_manager/test_classes.py [hunk 4]
```diff
@@ -46,6 +46,7 @@ def _create_jobs(self):
                 "file.exe",
                 "shellcode.bin",
                 "Sublime-Standard-Test-String.eml",
+                "textfile.txt",
```
Comment (mlodic) [inline tests/api_app/analyzers_manager/test_classes.py:49]: exactly, right, ty

**Review by mlodic — COMMENTED** [review]
thank you for the changes :)

File: api_app/analyzers_manager/file_analyzers/zippy_scan.py (hunk 1 above)
Comment (g4ze) [reply api_app/analyzers_manager/file_analyzers/zippy_scan.py:38]: noted

File: api_app/playbooks_manager/migrations/0026_add_zippy_scan_to_free_to_use.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+from django.db import migrations
+
+
+def migrate(apps, schema_editor):
+    playbook_config = apps.get_model("playbooks_manager", "PlaybookConfig")
+    AnalyzerConfig = apps.get_model("analyzers_manager", "AnalyzerConfig")
+    pc = playbook_config.objects.get(name="FREE_TO_USE_ANALYZERS")
+    pc.analyzers.add(AnalyzerConfig.objects.get(name="Mmdb_server").id)
```
Comment (mlodic) [inline api_app/playbooks_manager/migrations/0026_add_zippy_scan_to_free_to_use.py:12]: mmmm care about copy and pastes

Comment (g4ze) [reply api_app/playbooks_manager/migrations/0026_add_zippy_scan_to_free_to_use.py:12]: missed it by mistake; updating

**Comment by g4ze** [comment]
i think now everything works as intended.

**Review by mlodic — APPROVED** [review]

**Comment by mlodic** [comment]
great! merged!
